src/drivers/src/EncoderOdometryNode.cpp

```diff
@@ -0,0 +1,277 @@
+/*
+ * Created By: Gareth Ellis
+ * Created On: March 29th, 2018
+ * Description: A node the that translates the encoder readings (published in a
+ *              JointState message as # of ticks) to a ROS Odometry message
+ *              Originally designed for use with the Phidget 1047 encoder
+ */
+
+// ROS Includes
+#include <tf/transform_datatypes.h>
+
+// Snowbots Includes
+#include "EncoderOdometryNode.h"
+#include <sb_utils.h>
+
+
+EncoderOdometryNode::EncoderOdometryNode(int argc, char **argv, std::string node_name):
+    // Set current encoder ticks to "empty" as we haven't received any yet
+    left_encoder_num_ticks_curr(std::experimental::optional<int>{}),
+    right_encoder_num_ticks_curr(std::experimental::optional<int>{}),
+    left_encoder_num_ticks_prev(std::experimental::optional<int>{}),
+    right_encoder_num_ticks_prev(std::experimental::optional<int>{}),
+    last_estimate(generateAllZeroOdometryMessage())
+{
+    // Setup NodeHandles
+    ros::init(argc, argv, node_name);
+    ros::NodeHandle nh;
+    ros::NodeHandle private_nh("~");
+
+    // Get Params
+    SB_getParam(private_nh, "left_encoder_joint_name", left_encoder_joint_name, std::string("left_encoder"));
+    SB_getParam(private_nh, "right_encoder_joint_name", right_encoder_joint_name, std::string("right_encoder"));
+    SB_getParam(private_nh, "reverse_left_encoder", reverse_left_encoder, false);
+    SB_getParam(private_nh, "reverse_right_encoder", reverse_right_encoder, false);
+    SB_getParam(private_nh, "wheel_radius", wheel_radius, 0.1);
+    SB_getParam(private_nh, "wheelbase_length", wheelbase_length, 0.7);
+    SB_getParam(private_nh, "ticks_per_rotation", ticks_per_rotation, 1024);
+    SB_getParam(private_nh, "odom_frame", odom_frame_id, std::string("encoder"));
+    // TODO: Figure out actually reasonable variances
+    SB_getParam(private_nh, "left_encoder_variance", left_encoder_variance, 0.1);
+    SB_getParam(private_nh, "right_encoder_variance", right_encoder_variance, 0.1);
+
+    // Setup Subscriber(s)
+    joint_state_subscriber = nh.subscribe(std::string("/encoders/joint_states"), 10, &EncoderOdometryNode::encoderJointStateCallback, this);
+    reset_subscriber = private_nh.subscribe(std::string("reset"), 1, &EncoderOdometryNode::resetCallback, this);
+
+    // Setup Publisher(s)
+    std::string odom_estimate_topic_name = nh.resolveName("/encoders/odom");
```

Comment:
That was the intent here. I thought it might be a bit cleaner if all the encoder "stuff" (joint_states, Odometry), was under the same namespace. I've added an extra note to the code for now, but let me know if you think we ought to put it under a local namespace.